Model/Service/Sync.php

```diff
@@ -0,0 +1,261 @@
+use Nosto\Tagging\Model\ResourceModel\Product\Index\Collection as NostoIndexCollection;
+use Nosto\Tagging\Model\ResourceModel\Product\Index\CollectionFactory as NostoIndexCollectionFactory;
+use Nosto\Tagging\Util\Iterator;
+
+class Sync extends AbstractService
+{
+    const API_BATCH_SIZE = 50;
+    const PRODUCT_DELETION_BATCH_SIZE = 100;
+    const BENCHMARK_SYNC_NAME = 'nosto_product_sync';
+    const BENCHMARK_SYNC_BREAKPOINT = 1;
+    const BENCHMARK_DELETE_NAME = 'nosto_product_delete';
+    const BENCHMARK_DELETE_BREAKPOINT = 1;
+    const RESPONSE_TIMEOUT = 60;
+
+    /** @var IndexRepository */
+    private $indexRepository;
+
+    /** @var NostoHelperAccount */
+    private $nostoHelperAccount;
+
+    /** @var NostoHelperUrl */
+    private $nostoHelperUrl;
+
+    /** @var NostoIndexCollectionFactory */
+    private $nostoIndexCollectionFactory;
+
+    /** @var NostoDataHelper */
+    private $nostoDataHelper;
+
+    /**
+     * Index constructor.
+     * @param IndexRepository $indexRepository
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperUrl $nostoHelperUrl
+     * @param NostoLogger $logger
+     * @param NostoIndexCollectionFactory $nostoIndexCollectionFactory
+     * @param NostoDataHelper $nostoDataHelper
+     */
+    public function __construct(
+        IndexRepository $indexRepository,
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperUrl $nostoHelperUrl,
+        NostoLogger $logger,
+        NostoIndexCollectionFactory $nostoIndexCollectionFactory,
+        NostoDataHelper $nostoDataHelper
+    ) {
+        parent::__construct($nostoDataHelper, $logger);
+        $this->indexRepository = $indexRepository;
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperUrl = $nostoHelperUrl;
+        $this->nostoIndexCollectionFactory = $nostoIndexCollectionFactory;
+        $this->nostoDataHelper = $nostoDataHelper;
+    }
+
+    /**
+     * Handles sync of product collection by sending it to Nosto
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     * @throws MemoryOutOfBoundsException
+     */
+    public function syncIndexedProducts(NostoIndexCollection $collection, Store $store)
+    {
+        if (!$this->nostoDataHelper->isProductUpdatesEnabled($store)) {
+            $this->getLogger()->info(
+                'Nosto product sync is disabled - skipping upserting products to Nosto'
+            );
+            return;
+        }
+        $account = $this->nostoHelperAccount->findAccount($store);
+        if ($account instanceof NostoSignupAccount === false) {
+            throw new NostoException(sprintf('Store view %s does not have Nosto installed', $store->getName()));
+        }
+        $this->startBenchmark(self::BENCHMARK_SYNC_NAME, self::BENCHMARK_SYNC_BREAKPOINT);
+        $collection->setPageSize(self::API_BATCH_SIZE);
```

Comment:
Extract this bit of code into two chunks. The first method should be to sync the products and the next method should be to purge and delete the products. This is to increase readability. The resultant method should resemble this:


```
        try {
            $this->upsertProducts($store);
            $this->getLogger()->info(
                sprintf(
                    'Upserted products from index for store %s',
                    $store->getCode()
                )
            );
        } catch (NostoException $e) {
            $this->getLogger()->exception($e);
        }
        try {
            $this->purgeDeletedProducts($store);
            $this->getLogger()->info(
                sprintf(
                    'Removed products from index for store %s',
                    $store->getCode()
                )
            );
        } catch (NostoException $e) {
            $this->getLogger()->exception($e);
        }
```

Reply:
Done in `indexProducts()` method. https://github.com/Nosto/nosto-magento2/pull/496/files#diff-bc93a415469966c9efbe59308c5ad4dcR189

Reply:
Gotcha.